### PR TITLE
Add new cop Performance/Sample

### PR DESCRIPTION
**Pull request by rrosenblum**

### CODE REVIEW

File: lib/rubocop/cop/performance/sample.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      class Sample < Cop
+        MSG = 'Use `sample` instead of `shuffle.%s`.'
+        MSG_SELECTOR = 'Use `sample` instead of `shuffle[%s]`.'
+        ARRAY_SELECTORS = [:first, :last, :[]]
+
+        def on_send(node)
+          receiver, second_method, selector = *node
+          return unless ARRAY_SELECTORS.include? second_method
+          return if receiver.nil?
+          _, first_method = *receiver
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/sample.rb]: Avoid the uses of `_`.

File: lib/rubocop/cop/performance/sample.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      #   # good
+      #   [1, 2, 3].sample
+      class Sample < Cop
+        MSG = 'Use `sample` instead of `shuffle.%s`.'
+        MSG_SELECTOR = 'Use `sample` instead of `shuffle[%s]`.'
+        ARRAY_SELECTORS = [:first, :last, :[]]
+
+        def on_send(node)
+          receiver, second_method, selector = *node
+          return unless ARRAY_SELECTORS.include? second_method
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/sample.rb]: Add parens for `include?`.

File: lib/rubocop/cop/performance/sample.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                      format(MSG_SELECTOR, selector.loc.expression.source)
+                    else
+                      format(MSG, second_method)
+                    end
+
+          add_offense(node, range, message)
+        end
+
+        def autocorrect(node)
+          receiver, _ = *node
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/sample.rb]: You can drop `_` here.

File: spec/rubocop/cop/performance/sample_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  shared_examples 'register_offenses' do |selector|
+    it "when using shuffle#{selector} on an explicit array" do
+      inspect_source(cop, "[1, 2, 3, 4].shuffle#{selector}")
+
+      expect(cop.messages)
+        .to eq(["Use `sample` instead of `shuffle#{selector}`."])
+    end
+
+    it "when using shuffle#{selector} on an array that is set to a variable" do
```
Comment (bbatsov) [inline spec/rubocop/cop/performance/sample_spec.rb]: set -> assigned

**Comment by rrosenblum** [comment]
The code has been updated and rebased.

> Add parens for `include?`.
 Avoid the uses of `_`.
 You can drop `_` here.

It sounds like there could be cops that could check for these offenses.

**Comment by bbatsov** [comment]
:+1: 

> Add parens for include?.
> Avoid the uses of _.
> You can drop _ here.

First is pretty hard to do right. There have been several attempts so far. See https://github.com/bbatsov/rubocop/issues/1045

Second is also somewhat hard as in some cases `_` makes sense. Although having a disabled cop which people can enable would be fine I guess.

Third is definitely doable.

**Pull request merged**